cloudinit/net/network_state.py

```diff
@@ -1061,62 +1046,53 @@ def ipv6_mask_to_net_prefix(mask):
     If 'mask' is an integer or string representation of one then
     int(mask) will be returned.
     """
+    return mask_to_net_prefix(mask, 128)
 
-    if isinstance(mask, int):
-        return mask
+
+def mask_to_net_prefix(mask, max_bits=None):
+    """Return the network prefix for the netmask provided.
+
+    Supports ipv4 or ipv6 network mask.
+    """
+    if not isinstance(mask, (int, str)):
+        raise TypeError("Invalid network mask '%s'" % mask)
+
+    # If mask is a string, try to convert to integer
     if isinstance(mask, str):
         try:
-            return int(mask)
+            mask = int(mask)
         except ValueError:
             pass
-    else:
-        raise TypeError("mask '%s' is not a string or int")
-
-    if ":" not in mask:
-        raise ValueError("mask '%s' does not have a ':'")
-
-    bitCount = [
-        0,
-        0x8000,
-        0xC000,
-        0xE000,
-        0xF000,
-        0xF800,
-        0xFC00,
-        0xFE00,
-        0xFF00,
-        0xFF80,
-        0xFFC0,
-        0xFFE0,
-        0xFFF0,
-        0xFFF8,
-        0xFFFC,
-        0xFFFE,
-        0xFFFF,
-    ]
-    prefix = 0
-    for word in mask.split(":"):
-        if not word or int(word, 16) == 0:
-            break
-        prefix += bitCount.index(int(word, 16))
 
-    return prefix
+    # If mask passed as integer, or convert to int succeded
+    if isinstance(mask, int):
+        if max_bits and not (0 <= mask <= max_bits):
```

Comment:
> I think that this logic is wrong. Here the converted integer value of an address is compared against the number of bits. I think this needs to be compared to `2^max_bits - 1` I don't see any tests that test the "int conversion" codepath, so I think it would be good to exercise this codepath either way.
> 
> If I understand this code path correctly, the intention is that "mask" passed in is an `IPv4Address` object, correct?
> 

No, this method will raise `TypeError` if a `IPv4Address` or `IPv6Address` is passed.
In `cloudinit/net/network_state.py:L1056` there is a condition on `str`or `int` instance, `IPv4Address` or `IPv6Address` is neither. However, we should add `IPv4Address` and `IPv6Address` to that condition so that these types are allowed.

The handling of `int` and `str` that can be converted to `int` here is to keep the functionality provided by the methods being refactored. The idea is that the caller does not know if the netmask it has is represented as a `prefix` or IPvXAddress string. In the case the value is already a `prefix` either `int(:number:)` for `str(:number:)`the methods will simply "echo" back the number as int.

> For reference:
> 
> ```
> >>> import ipaddress
> >>> 2**32 - 1 == int(ipaddress.IPv4Address("255.255.255.255"))
> True
> ```
> 

The `(0 <= mask <= max_bits)` logic is simply trying to make sure a `prefix` value passed is valid. e.g 0-32 for IPv4 and 0-128 for IPv6.

> Alternatively, if we aren't actually using this codepath anywhere (haven't checked) we could could just remove it.

The `mask_to_net_prefix` is used a bit:
```
cloudinit/net/__init__.py:from cloudinit.net.network_state import mask_to_net_prefix
cloudinit/net/__init__.py:            self.prefix = mask_to_net_prefix(prefix_or_mask)
cloudinit/net/network_state.py:            prefix = mask_to_net_prefix(maybe_prefix)
cloudinit/net/network_state.py:        prefix = mask_to_net_prefix(netmask)
cloudinit/net/network_state.py:    return mask_to_net_prefix(mask, 32)
cloudinit/net/network_state.py:    return mask_to_net_prefix(mask, 128)
cloudinit/net/network_state.py:def mask_to_net_prefix(mask, max_bits=None):
cloudinit/net/network_state.py:    mask_dec = ipv4_mask_to_net_prefix(mask)
cloudinit/net/sysconfig.py:                prefix_value = network_state.ipv6_mask_to_net_prefix(
cloudinit/sources/helpers/vmware/imc/config_nic.py:        cidr = mask_to_net_prefix(netmask)
cloudinit/sources/DataSourceOpenNebula.py:            prefix = str(net.mask_to_net_prefix(mask))
```

Reply:
> No, this method will raise `TypeError` if a `IPv4Address` or `IPv6Address` is passed. In `cloudinit/net/network_state.py:L1056` there is a condition on `str`or `int` instance, `IPv4Address` or `IPv6Address` is neither. 

I missed that, thanks.

> However, we should add `IPv4Address` and `IPv6Address` to that condition so that these types are allowed.
> The handling of `int` and `str` that can be converted to `int` here is to keep the functionality provided by the methods being refactored. The idea is that the caller does not know if the netmask it has is represented as a `prefix` or IPvXAddress string. In the case the value is already a `prefix` either `int(:number:)` for `str(:number:)`the methods will simply "echo" back the number as int.



> > Alternatively, if we aren't actually using this codepath anywhere (haven't checked) we could could just remove it.
> 
> The `mask_to_net_prefix` is used a bit:
> 
> ```
> cloudinit/net/__init__.py:from cloudinit.net.network_state import mask_to_net_prefix
> cloudinit/net/__init__.py:            self.prefix = mask_to_net_prefix(prefix_or_mask)
> cloudinit/net/network_state.py:            prefix = mask_to_net_prefix(maybe_prefix)
> cloudinit/net/network_state.py:        prefix = mask_to_net_prefix(netmask)
> cloudinit/net/network_state.py:    return mask_to_net_prefix(mask, 32)
> cloudinit/net/network_state.py:    return mask_to_net_prefix(mask, 128)
> cloudinit/net/network_state.py:def mask_to_net_prefix(mask, max_bits=None):
> cloudinit/net/network_state.py:    mask_dec = ipv4_mask_to_net_prefix(mask)
> cloudinit/net/sysconfig.py:                prefix_value = network_state.ipv6_mask_to_net_prefix(
> cloudinit/sources/helpers/vmware/imc/config_nic.py:        cidr = mask_to_net_prefix(netmask)
> cloudinit/sources/DataSourceOpenNebula.py:            prefix = str(net.mask_to_net_prefix(mask))
> ```

I know the function is used in several places, what I meant here is that if no calls to `mask_to_net_prefix()` with an integer type exist, then we could simplify the function by removing support for passing in non-strings. My reasoning is that, although functions that support multiple input types are convenient, they add debt in that they are more error prone / difficult to review / audit, so if we aren't using them it's better to simplify, that's all.

Reply:
> I know the function is used in several places, what I meant here is that if no calls to `mask_to_net_prefix()` with an integer type exist, then we could simplify the function by removing support for passing in non-strings. My reasoning is that, although functions that support multiple input types are convenient, they add debt in that they are more error prone / difficult to review / audit, so if we aren't using them it's better to simplify, that's all.

Ah, yes now I understand. And I fully agree!

If I remove `int` from the allowd types at `cloudinit/net/network_state.py:L1056` and run unit tests it looks like only two tests fail:
```
FAILED tests/unittests/test_net.py::TestEniRoundTrip::test_ipv6_static_routes - TypeError: Invalid network mask '24'
FAILED tests/unittests/net/test_init.py::TestEphemeralIPV4Network::test_ephemeral_ipv4_network_with_prefix - TypeError: Invalid network mask '16'
```

Might not be so much work to handle this at the caller side instead.

Consider deprecate non-string, log a warning and then follow up with a patch to remove support for passing non-string in a future release?

Reply:
> If I remove `int` from the allowd types at `cloudinit/net/network_state.py:L1056` and run unit tests it looks like only two tests fail:
> 
> ```
> FAILED tests/unittests/test_net.py::TestEniRoundTrip::test_ipv6_static_routes - TypeError: Invalid network mask '24'
> FAILED tests/unittests/net/test_init.py::TestEphemeralIPV4Network::test_ephemeral_ipv4_network_with_prefix - TypeError: Invalid network mask '16'
> ```
> Might not be so much work to handle this at the caller side instead.


I analyzed the Eni parser yesterday. It doesn't cast to an int anywhere. I think the parsed input from Eni will always be a string.
Since `test_ephemeral_ipv4_network_with_prefix()` seems to be generic with respect to datasource/renderer, I don't think that is suggesting that any parser requires using the integer argument, just exists for test coverage.


> Consider deprecate non-string, log a warning and then follow up with a patch to remove support for passing non-string in a future release?


Since this function is internal to cloud-init, I don't think that it makes much sense to log a deprecation warning. Users don't have anything actionable to do about it (they don't have control over the type I don't think) so it would just be noise. Since it appears that the integer argument codepath is executed purely by tests, I think it makes more sense to just rip it out.


One more realization I had yesterday while reviewing this is that all calls to `mask_to_net_prefix()` in the code appear to be in a context where the ip address version is actually known. This means that a generic `mask_to_net_prefix()` isn't really a hard requirement - just a convenience. Since the `ipv(4|6)_mask_to_net_prefix()` implementations always call `mask_to_net_prefix()` with a `max_bits` value, they provide better validation. With this in mind, I would argue that we should probably replace `mask_to_net_prefix()` with the `ipv(4|6)_` versions to get stronger validation, and perhaps change `mask_to_net_prefix()` to `_mask_to_net_prefix()`.

Reply:
I removed `mask_to_net_prefix()` completely and refactored `ipv(4|6)_mask_to_net_prefix` to use ipaddress from the library when possible. I updated code in the few places where  `mask_to_net_prefix()` was used directly to use the version specific methods.

AFICT, removing `mask_to_net_prefix()` turned out to be a lot less of a hassle than I thought it would be.

Thanks for suggesting it!

Reply:
@hjensas I think this approach looks much better, thanks!